File: backend/src/models/user.model.js
const { Schema, model } = require("mongoose");
const uuid = require("uuid");
const bcrypt = require("bcrypt");

const userSchema = new Schema(
  {
    id: {
      type: String,
      unique: true,
    },
    fullname: {
      type: String,
      default: "",
    },
    email: {
      type: String,
      unique: true,
      required: true,
    },
    password: {
      type: String,
      required: true,
    },
    phone: {
      type: String,
      default: "",
    },
    city: {
      type: String,
      default: "",
    },
    state: {
      type: String,
      default: "",
    },
    address: {
      type: String,
      default: "",
    },
    profileProgress: {
      type: Number,
      default: 0, // 0 -->account created, 1 --> address and details entered
    },
    updatedOn: {
      type: Date,
    },
    createdOn: {
      type: Date,
    },
  },
  { timestamps: true }
);
userSchema.pre("save", function (next) {
  this.id = uuid.v1();
  this.updatedOn = new Date();
  this.createdOn = new Date();

  // hashing the password
  const salt = bcrypt.genSaltSync(10);
  const hash = bcrypt.hashSync(this.password, salt); // hash -->hashed password
  this.password = hash;

  next();
});
userSchema.pre(["update", "findOneAndUpdate", "updateOne"], function (next) {
  // _id -->mongodb object id
  const update = this.getUpdate();
  delete update._id;
  delete update.id;

  this.updatedOn = new Date();
  next();
});

const UserModel = model("User", userSchema);
module.exports = UserModel;

File: frontend/lib/data/models/user/user.model.dart
class userModel {
  String? sId;
  String? fullname;
  String? email;
  String? password;
  String? phone;
  String? city;
  String? state;
  String? address;
  int? profileProgress;
  String? createdAt;
  String? updatedAt;
  String? id;
  String? updatedOn;
  String? createdOn;

  userModel(
      {this.sId,
      this.fullname,
      this.email,
      this.password,
      this.phone,
      this.city,
      this.state,
      this.address,
      this.profileProgress,
      this.createdAt,
      this.updatedAt,
      this.id,
      this.updatedOn,
      this.createdOn});

  userModel.fromJson(Map<String, dynamic> json) {
    sId = json['_id'];
    fullname = json['fullname'];
    email = json['email'];
    password = json['password'];
    phone = json['phone'];
    city = json['city'];
    state = json['state'];
    address = json['address'];
    profileProgress = json['profileProgress'];
    createdAt = json['createdAt'];
    updatedAt = json['updatedAt'];
    id = json['id'];
    updatedOn = json['updatedOn'];
    createdOn = json['createdOn'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['_id'] = this.sId;
    data['fullname'] = this.fullname;
    data['email'] = this.email;
    data['password'] = this.password;
    data['phone'] = this.phone;
    data['city'] = this.city;
    data['state'] = this.state;
    data['address'] = this.address;
    data['profileProgress'] = this.profileProgress;
    data['createdAt'] = this.createdAt;
    data['updatedAt'] = this.updatedAt;
    data['id'] = this.id;
    data['updatedOn'] = this.updatedOn;
    data['createdOn'] = this.createdOn;
    return data;
  }
}
